daliuge-engine/dlg/io.py

```diff
@@ -424,25 +428,78 @@ def __init__(self, object_id, plasma_path='/tmp/plasma'):
         super(PlasmaIO, self).__init__()
         self._plasma_path = plasma_path
         self._object_id = object_id
+        self._reader = None
+        self._writer = None
 
     def _open(self, **kwargs):
         return plasma.connect(self._plasma_path)
 
     def _close(self, **kwargs):
+        if self._writer:
+            self._desc.put_raw_buffer(self._writer.getvalue(), self._object_id)
+            self._writer.close()
+        if self._reader:
+            self._reader.close()
+
+    def _read(self, count, **kwargs):
+        if not self._reader:
+            [data] = self._desc.get_buffers([self._object_id])
+            self._reader = pyarrow.BufferReader(data)
+        return self._reader.read1(count)
+
+    def _write(self, data, **kwargs):
+        if not self._writer:
+            # use client.create and FixedSizeBufferWriter
+            self._writer = pyarrow.BufferOutputStream()
+        self._writer.write(data)
```

Comment:
I rewrote the plasma writing here as the DataIO interface expects to write the number of bytes in `count` and calls `write` multiple times. Tested with a large file and this now works

Reply:
To support the direct writing optimization it should be possible to support a fixed-drop-size mode where a drop-size parameter is used and an error is thrown if the size when the stream ends does not match the plasma store buffer (but allows the developer to update the graph with the size printed in the error). This of course would be an optional feature since some programs generate varying sized outputs

Reply:
Fixed sized buffer writer is now used